### PR TITLE
Added `locations` field to VLAN.

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #4412 
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/ipam/forms.py [hunk 1]
```diff
@@ -682,6 +682,13 @@ class VLANGroupFilterForm(NautobotFilterForm, LocatableModelFilterFormMixin):
 
 
 class VLANForm(LocatableModelFormMixin, NautobotModelForm, TenancyForm):
+    locations = DynamicModelMultipleChoiceField(
+        queryset=Location.objects.all(),
+        required=False,
+        label="Locations",
+        null_option="None",
+        # query_params={"location": "$location"},
```
Comment (timizuoebideri1) [inline nautobot/ipam/forms.py:690]:
```suggestion
```

**Review by glennmatthews — COMMENTED** [review]
Feels like it'll be confusing to have both `.location` and `.locations` on the VLAN model, in the VLANForm, in the detail view etc. I get that backwards compatibility is a factor but I wonder if there's a more elegant way we can do this (e.g. changing `VLAN.location` from a concrete DB field to an `@property` or similar?)

Also, doesn't look like this addresses the VLAN group location constraints in `VLAN.clean()` - per the comments on the parent issue I think we were considering removing those?

**Comment by gsnider2195** [comment]
@timizuoebideri1 lets move this PR back to draft until we can have a design discussion

File: nautobot/ipam/api/views.py [hunk 2]
```diff
@@ -24,6 +25,7 @@
     VRF,
 )
 from . import serializers
+from ...dcim.models import Location
```
Comment (gsnider2195) [inline nautobot/ipam/api/views.py:28]:
```suggestion
from nautobot.dcim.models import Location
```

File: nautobot/ipam/forms.py [hunk 3]
```diff
@@ -703,23 +712,40 @@ class Meta:
             "tags",
         ]
         help_texts = {
-            "location": "Leave blank if this VLAN spans multiple locations",
+            "locations": "Leave blank if this VLAN spans multiple locations",
```
Comment (glennmatthews) [inline nautobot/ipam/forms.py:715]: Perhaps:

```suggestion
            "locations": "Leave blank if this VLAN spans all locations",
```

File: nautobot/ipam/forms.py [hunk 4]
```diff
@@ -703,23 +712,40 @@ class Meta:
...
 
 class VLANBulkEditForm(
     TagsBulkEditFormMixin,
-    LocatableModelBulkEditFormMixin,
     StatusModelBulkEditFormMixin,
     RoleModelBulkEditFormMixin,
     NautobotBulkEditForm,
 ):
     pk = forms.ModelMultipleChoiceField(queryset=VLAN.objects.all(), widget=forms.MultipleHiddenInput())
+    # TODO (timizuo): Should we have locations in VLAN bulk create
```
Comment (glennmatthews) [inline nautobot/ipam/forms.py:739]: In bulk edit, it'd be appropriate to have `locations_add` and `locations_remove` fields, similar to how we do in `TagsBulkEditFormMixin`.

File: nautobot/ipam/models.py [hunk 5]
```diff
@@ -1370,28 +1368,43 @@ class Meta:
...
+
+    @location.setter
+    def location(self, value):
+        if self.locations.count() > 1:
+            raise self.locations.model.MultipleObjectsReturned(
+                "Multiple Location objects returned. Please refer to locations."
             )
+        self.locations.set([value])
+
+    def location_ancestors(self, include_self=False):
```
Comment (glennmatthews) [inline nautobot/ipam/models.py:1400]: Unused?

File: nautobot/ipam/tables.py [hunk 6]
```diff
@@ -688,7 +688,6 @@ class Meta(BaseTable.Meta):
 class VLANTable(StatusTableMixin, RoleTableMixin, BaseTable):
     pk = ToggleColumn()
     vid = tables.TemplateColumn(template_code=VLAN_LINK, verbose_name="ID")
-    location = tables.Column(linkify=True)
```
Comment (glennmatthews) [inline nautobot/ipam/tables.py:691]: As mentioned above, should probably add a `locations` column as a `LinkedCountColumn` here and below.

File: nautobot/ipam/templates/ipam/vlan_edit.html [hunk 7]
```diff
@@ -16,6 +16,7 @@
         <div class="panel-heading"><strong>Assignment</strong></div>
         <div class="panel-body">
             {% render_field form.location %}
```
Comment (glennmatthews) [inline nautobot/ipam/templates/ipam/vlan_edit.html:18]: remove?

File: nautobot/ipam/tests/test_filters.py [hunk 8]
```diff
@@ -1013,7 +1013,6 @@ def setUpTestData(cls):
             VLAN.objects.create(
                 vid=101,
                 name="VLAN 101",
-                location=cls.locations[0],
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_filters.py:1016]: This should still work with the way you've implemented `VLAN.__init__` and `save` right?

File: nautobot/ipam/tests/test_filters.py [hunk 9]
```diff
@@ -1078,14 +1079,15 @@ def test_vid(self):
         params = {"vid": vids}
         self.assertQuerysetEqual(self.filterset(params, self.queryset).qs, self.queryset.filter(vid__in=vids))
 
-    def test_location(self):
-        params = {"location": [self.locations[0].pk, self.locations[1].pk]}
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_filters.py:1082]: Need to still support and test a `location` filter.

File: nautobot/ipam/views.py [hunk 10]
```diff
@@ -1226,7 +1226,7 @@ def get_extra_context(self, request, instance):
         vlan_table = tables.VLANDetailTable(vlans)
         if request.user.has_perm("ipam.change_vlan") or request.user.has_perm("ipam.delete_vlan"):
             vlan_table.columns.show("pk")
-        vlan_table.columns.hide("location")
+        # vlan_table.columns.hide("location")
```
Comment (glennmatthews) [inline nautobot/ipam/views.py:1229]: delete? Or need to hide `locations` column instead?

File: nautobot/ipam/views.py [hunk 11]
```diff
@@ -1277,7 +1277,7 @@ class VLANGroupBulkDeleteView(generic.BulkDeleteView):
 
 
 class VLANListView(generic.ObjectListView):
-    queryset = VLAN.objects.select_related("location", "vlan_group", "tenant", "role", "status")
+    queryset = VLAN.objects.select_related("vlan_group", "tenant", "role", "status")
```
Comment (glennmatthews) [inline nautobot/ipam/views.py:1280]: Probably need a `count_related` here and below once you add the `locations` `LinkedCountColumn` to the table.

File: nautobot/dcim/api/views.py [hunk 12]
```diff
@@ -154,7 +154,7 @@ class LocationViewSet(NautobotModelViewSet):
             device_count=count_related(Device, "location"),
             rack_count=count_related(Rack, "location"),
             prefix_count=count_related(Prefix, "location"),
-            vlan_count=count_related(VLAN, "location"),
+            vlan_count=count_related(VLAN, "locations"),
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/dcim/api/views.py:159]: Yup

File: nautobot/ipam/tests/test_filters.py (hunk 8 above)
Comment (timizuoebideri1) [reply nautobot/ipam/tests/test_filters.py:1016]: Yup

File: changes/4412.added [hunk 13]
```diff
@@ -0,0 +1 @@
+Added `locations` field to VLAN.
```
Comment (glennmatthews) [inline changes/4412.added:1]: Also should be documented in the `VLAN` model documentation and summarized in the 2.2 release notes release overview.

File: nautobot/ipam/api/serializers.py [hunk 14]
```diff
@@ -135,6 +135,37 @@ def validate(self, data):
...
+            "last_updated",
+            "tags",
+            "notes_url",
+            "custom_fields",
+        ]
+        validators = []
+
+    def update(self, instance, validated_data):
+        print("VALIDATED DATA ====> ", validated_data)
+        return super().update(instance, validated_data)
```
Comment (glennmatthews) [inline nautobot/ipam/api/serializers.py:166]: Remove?

File: nautobot/ipam/api/views.py [hunk 15]
```diff
@@ -311,11 +313,19 @@ class VLANGroupViewSet(NautobotModelViewSet):
...
 
 
+@extend_schema_view(
+    bulk_update=extend_schema(responses={"200": serializers.VLANLegacySerializer(many=True)}, versions=["2.0"]),
+    bulk_partial_update=extend_schema(responses={"200": serializers.VLANLegacySerializer(many=True)}, versions=["2.0"]),
+    create=extend_schema(responses={"201": serializers.VLANLegacySerializer}, versions=["2.0"]),
+    list=extend_schema(responses={"200": serializers.VLANLegacySerializer(many=True)}, versions=["2.0"]),
+    partial_update=extend_schema(responses={"200": serializers.VLANLegacySerializer}, versions=["2.0"]),
+    retrieve=extend_schema(responses={"200": serializers.VLANLegacySerializer}, versions=["2.0"]),
+    update=extend_schema(responses={"200": serializers.VLANLegacySerializer}, versions=["2.0"]),
```
Comment (glennmatthews) [inline nautobot/ipam/api/views.py:323]: all of these should be `["2.0", "2.1"]` right?

File: nautobot/ipam/tables.py [hunk 16]
```diff
@@ -733,7 +735,6 @@ class Meta(VLANTable.Meta):
         default_columns = (
             "pk",
             "vid",
-            "location",
```
Comment (glennmatthews) [inline nautobot/ipam/tables.py:736]: Should this table also have a `location_count` column added?

File: nautobot/ipam/views.py [hunk 17]
```diff
@@ -1291,7 +1292,6 @@ class VLANListView(generic.ObjectListView):
 class VLANView(generic.ObjectView):
     queryset = VLAN.objects.select_related(
         "role",
-        "location",
```
Comment (glennmatthews) [inline nautobot/ipam/views.py:1294]: Do we need a `location_count` here?

File: nautobot/ipam/models.py [hunk 18]
```diff
@@ -1342,7 +1341,7 @@ class VLAN(PrimaryModel):
     description = models.CharField(max_length=200, blank=True)
 
     clone_fields = [
-        "location",
+        "locations",
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/ipam/models.py:1344]: yup, works good

File: nautobot/core/tests/test_api.py [hunk 19]
```diff
@@ -425,13 +425,13 @@ def test_serializer_class_required(self):
...
-        location = dcim_models.Location.objects.filter(parent__isnull=False).first()
+        vlan_group = ipam_models.VLANGroup.objects.filter(location__isnull=False).first()
 
         csv_data = "\n".join(
             [
                 # the "foobar" column is not understood by the serializer; per usual REST API behavior, it'll be skipped
-                "vid,name,foobar,description,status,tags,tenant,location__name,location__parent__name",
-                f'22,hello,huh?,"It, I say, is a living!",{status.name},"{tags.first().name},{tags.last().name}",,{location.name},{location.parent.name}',
+                "vid,name,foobar,description,status,tags,tenant,vlan_group__name,vlan_group__location__name",
+                f'22,hello,huh?,"It, I say, is a living!",{status.name},"{tags.first().name},{tags.last().name}",,{vlan_group.name},{vlan_group.location.name}',
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot/core/tests/test_api.py:431]: I believe so, since location is not a real model field anymore

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/core/tests/test_api.py:431]: Since CSV import is REST serializer based now, I wonder if it'd be worth the effort to add `location` as an optional write-only field on the "new" serializer and make it be able to handle it?

File: nautobot/docs/user-guide/core-data-model/ipam/vlan.md [hunk 20]
```diff
@@ -15,3 +15,6 @@ As with prefixes, each VLAN may also be assigned a functional role. Prefixes and
 
 +/- 2.0.0
     - Renamed `group` field to `vlan_group`.
+
++/- 2.2.0
+    - Replace `location` ForeginKey field with `locations` ManyToManyField
```
Comment (glennmatthews) [inline nautobot/docs/user-guide/core-data-model/ipam/vlan.md:20]:
```suggestion
    - Replaced `location` ForeignKey field with `locations` ManyToManyField, allowing a VLAN to be assigned to multiple Locations.
```

File: nautobot/core/testing/filters.py [hunk 21]
```diff
@@ -41,7 +41,7 @@ def get_filterset_test_values(self, field_name, queryset=None):
                 # randomly break out of loop after 2 values have been selected
                 if len(test_values) > 1 and random.choice([True, False]):  # noqa: S311  # suspicious-non-cryptographic-random-usage
                     break
-                if value[field_name] and value["count"] < qs_count:
+                if value[field_name] and (len(test_values) < 2 or value["count"] < qs_count):
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:44]: I think this makes it so that we always pick the first two values that we find, rather than random ones - is that change intentional/necessary?

File: nautobot/core/views/generic.py [hunk 22]
```diff
@@ -940,6 +940,9 @@ def alter_obj(self, obj, request, url_args, url_kwargs):
         # For example, a parent object can be defined given some parameter from the request URL.
         return obj
 
+    def _extra_post_save_action(self, obj, form):
```
Comment (glennmatthews) [inline nautobot/core/views/generic.py:943]: Nit: since this is intended for views to override, I'm not sure that the `_` prefix is necessary/appropriate.

File: nautobot/dcim/tests/test_filters.py [hunk 23]
```diff
@@ -670,6 +670,11 @@ class LocationFilterSetTestCase(FilterTestCases.NameOnlyFilterTestCase, FilterTe
     def setUpTestData(cls):
         common_test_data(cls)
 
+        # Since VLANFactory is not smart enough to create some VLANs without locations we have to manually remove locations form some vlans
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_filters.py:673]: Is this still true/necessary after changing the `minimum` count to 0 in VLANFactory?

File: nautobot/ipam/api/serializers.py [hunk 24]
```diff
@@ -135,6 +136,37 @@ def validate(self, data):
         return data
 
 
+class VLANLegacySerializer(VLANSerializer):
```
Comment (glennmatthews) [inline nautobot/ipam/api/serializers.py:139]: Mind adding a docstring spelling out that this is used for API versions 2.0 and 2.1?

File: nautobot/ipam/tests/test_api.py [hunk 25]
```diff
@@ -741,6 +748,62 @@ def test_delete_vlan_with_prefix(self):
...
+            data["location"] = data.pop("locations")[0]
+            data.pop("vlan_group")
+            response = self.client.patch(f"{url}?api_version=2.1", data, format="json", **self.header)
+            self.assertHttpStatus(response, status.HTTP_412_PRECONDITION_FAILED)
+            self.assertEqual(
+                str(response.data["detail"]),
+                "This object does not conform to pre-2.2 behavior. Please correct data or use API version 2.2",
+            )
+
+        # TODO (timizuo): Fix Update is not currently updating location
```
Comment (glennmatthews) [inline nautobot/ipam/tests/test_api.py:796]: Any idea what the issue may be here?

File: nautobot/virtualization/forms.py [hunk 26]
```diff
@@ -577,7 +577,7 @@ class VMInterfaceBulkEditForm(TagsBulkEditFormMixin, StatusModelBulkEditFormMixi
         queryset=VLAN.objects.all(),
         required=False,
         query_params={
-            "location_id": "null",
+            "locations": "null",
```
Comment (glennmatthews) [inline nautobot/virtualization/forms.py:580]: Also needed at line 573, I think?

File: nautobot/core/testing/filters.py (hunk 21 above)
Comment (HanlinMiao) [reply nautobot/core/testing/filters.py:44]: revert it for now see what breaks

File: nautobot/dcim/tests/test_filters.py (hunk 23 above)
Comment (HanlinMiao) [reply nautobot/dcim/tests/test_filters.py:673]: Still necessary, but I took an alternative route. Instead of clearing the vlans here, I disallowed vlan content type on `Building` and `Floor` LocationType locations.

File: nautobot/ipam/tests/test_api.py (hunk 25 above)
Comment (HanlinMiao) [reply nautobot/ipam/tests/test_api.py:796]: Seems like there is no issue after I fixed the attribute error.

File: nautobot/core/tests/test_api.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/core/tests/test_api.py:431]: Added location field as an optional write-only field. But should we revert the test changes here?

File: nautobot/core/testing/filters.py [hunk 27]
```diff
@@ -41,7 +41,7 @@ def get_filterset_test_values(self, field_name, queryset=None):
                 # randomly break out of loop after 2 values have been selected
                 if len(test_values) > 1 and random.choice([True, False]):  # noqa: S311  # suspicious-non-cryptographic-random-usage
                     break
-                if value[field_name] and value["count"] < qs_count:
+                if value[field_name] and value["count"] <= qs_count:
```
Comment (HanlinMiao) [inline nautobot/core/testing/filters.py:44]: Change made here to account for scenarios when there are exactly two choices.

**Review by glennmatthews — APPROVED** [review]

File: nautobot/core/testing/filters.py (hunk 27 above)
Comment (gsnider2195) [reply nautobot/core/testing/filters.py:44]: This test should fail when there are exactly two choices. You need more than two choices to accurately test a multiplechoicefilter
